sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -713,13 +722,38 @@ ProgramManager::collectDeviceImageDepsForImportedSymbols(
     bool Found = false;
     for (auto It = Range.first; It != Range.second; ++It) {
       RTDeviceBinaryImage *Img = It->second;
-      if (Img->getFormat() != Format ||
-          !doesDevSupportDeviceRequirements(Dev, *Img) ||
+
+      if (!doesDevSupportDeviceRequirements(Dev, *Img) ||
           !compatibleWithDevice(Img, *getSyclObjImpl(Dev).get()))
         continue;
+
+      // If the image is a special device image, we need to check if it
+      // should be used for this device.
       if (isSpecialDeviceImage(Img) &&
           !isSpecialDeviceImageShouldBeUsed(Img, *getSyclObjImpl(Dev).get()))
         continue;
+
+      // Check if the image format matches the main image format.
+      // If any of the images is compressed, we need to decompress it
+      // and then check if the format matches.
+      if (Img->getFormat() != Format) {
+        if (Format == SYCL_DEVICE_BINARY_TYPE_COMPRESSED_NONE ||
+            Img->getFormat() == SYCL_DEVICE_BINARY_TYPE_COMPRESSED_NONE) {
+
+          auto MainImgPtr = const_cast<RTDeviceBinaryImage *>(&MainImg);
+          CheckAndDecompressImage(MainImgPtr);
+          CheckAndDecompressImage(Img);
+          Format = MainImg.getFormat();
+
+          // Nope, we can't link two images with different formats.
+          if (Img->getFormat() != Format)
+            continue;
+        }
+        // Skip this image as its format differs from the main image.
+        else
+          continue;
+      }
```

Comment:
What happens if both are compressed? Do we just continue on linking the compressed images? Otherwise, we could:

```suggestion
      if (Format == SYCL_DEVICE_BINARY_TYPE_COMPRESSED_NONE ||
          Img->getFormat() == SYCL_DEVICE_BINARY_TYPE_COMPRESSED_NONE) {
        auto MainImgPtr = const_cast<RTDeviceBinaryImage *>(&MainImg);
        CheckAndDecompressImage(MainImgPtr);
        CheckAndDecompressImage(Img);
        Format = MainImg.getFormat();
      }
      // Skip this image if its format differs from the main image.
      if (Img->getFormat() != Format)
        continue;
```